### ISSUE TITLE
improve dzsave performance

### ISSUE THREAD

**Issue by jcupitt** [libvips/libvips#3067]
With this test image:

https://openslide.cs.cmu.edu/download/openslide-testdata/Aperio/CMU-1.svs

And git master libvips I see:

```
$ time vips dzsave CMU-1.svs x
memory: high-water mark 468.29 MB
real	1m1.608s
user	6m12.874s
sys	16m7.571s
```

CPU utilization is not great, and systime is extremely high.

Using this:

https://github.com/openslide/openslide-python/blob/main/examples/deepzoom/deepzoom_tile.py

I see:

```
$ time ./deepzoom_tile.py ~/pics/openslide/CMU-1.svs -o x -j 32
Tiling slide: wrote 31644/31644 tiles
real	0m10.273s
user	5m8.292s
sys	0m14.442s
```

So 6x faster.

**Comment by jcupitt** [comment]
I comment out the call to `write_image()` in dzsave I see:

```
$ time vips dzsave CMU-1.svs x
real	0m9.291s
user	1m6.651s
sys	0m9.174s
```

That's the time to scan the SVS and generate every tile in every layer, so the other 50s is being spent writing the JPG tiles.

**Comment by kleisauke** [comment]
Comment https://github.com/libvips/libvips/discussions/2038#discussioncomment-377004 might also be relevant here. Given this is an whole-slide image, we could also experiment with having `openslideload` output flattened RGB rather than RGBA to avoid the `vips_flatten` at write.

**Comment by jcupitt** [comment]
Yes, I remember that discussion. I should have spent more time looking into this back then sigh.

I tried instrumenting `vips__thread_execute()` and it's peaking at 1086 active threads in the threadpool :( I wonder if GThreadpool doesn't scale well to high thread counts? It's doing a huge amount of locking and unlocking.

**Comment by jcupitt** [comment]
> we could also experiment with having openslideload output flattened RGB rather than RGBA to avoid the vips_flatten at write.

You're right, we could try adding something like an `rgb` flag to `openslideload`. I don't know if it would make a huge difference to runtime.

**Comment by kleisauke** [comment]
I see this on my 16-core workstation:

```bash
$ time vips dzsave CMU-1.svs x

real	0m35.642s
user	1m51.041s
sys	0m30.529s

# glibc >= 2.34 uses the clone3 syscall instead
$ strace -cfe trace=clone3 vips dzsave CMU-1.svs x
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
100.00    0.004959          79        62           clone3
------ ----------- ----------- --------- --------- ----------------
100.00    0.004959          79        62           total

$ vips flatten CMU-1.svs x.tiff[tile]
$ vipsheader x.tiff
x.tiff: 46000x32914 uchar, 3 bands, srgb, tiffload
$ time vips dzsave x.tiff x2

real	0m19.330s
user	0m50.595s
sys	0m30.506s
$ sha256sum x{,2}_files/16/0_0.jpeg
e2f84d7856781d93df6e7e0317b1426091ae70c72f3dd9b2d6059b9b4e524f2d  x_files/16/0_0.jpeg
e2f84d7856781d93df6e7e0317b1426091ae70c72f3dd9b2d6059b9b4e524f2d  x2_files/16/0_0.jpeg
```

So, it uses 62(!) active threads here, and flattening it to a TIFF image prior the dzsave makes it ~45% faster (but that's probably not a fair comparison).

> I wonder if GThreadpool doesn't scale well to high thread counts? It's doing a huge amount of locking and unlocking.

Indeed, I also noticed this when `strace`-ing the futex syscall. We could try to limit the number of threads running concurrently in the threadpool (it's unlimited now), but I'm not sure if that would fix this.

**Comment by jcupitt** [comment]
> So, it uses 62(!) active threads here, and flattening it to a TIFF image prior the dzsave makes it ~45% faster (but that's probably not a fair comparison).

I think openslide is going via a cairo surface, so it's relatively slow. The TIFF loader in master is quite a bit quicker.

I'll try quickly implementing a threadpool class again -- we don't need a lot of the locking that GThreadopool does, since vips workers already queue up to allocate work.

**Comment by jcupitt** [comment]
>  So, it uses 62(!) active threads here,

That's interesting, I turned on this thing in `vips__thread_execute()` and I see over 1000 threads in the pool:
 
```C
         VIPS_DEBUG_MSG( "vips__thread_execute: %u threads in pool\n",
                g_thread_pool_get_num_threads( vips__pool ) );
```

Perhaps glib is limiting the number of threads used to service the queue?

**Comment by jcupitt** [comment]
I reimplemented GThreadpool in the slimmest way I could:

https://github.com/libvips/libvips/blob/reimplement-threadpool/libvips/iofuncs/threadset.c

I now see:

```
$ time vips dzsave CMU-1.svs x
real	1m0.311s
user	5m22.432s
sys	17m58.646s
```

So it's perhaps a little quicker, but not by a useful amount lol.

Something else is going wrong.

**Comment by kleisauke** [comment]
> I turned on this thing in `vips__thread_execute()` and I see over 1000 threads in the pool

Ah, I could reproduce that with `--vips-concurrency 64`. Doing `--vips-concurrency 6` (which corresponds to the number of cores on my i5-8600K CPU), doesn't consume that many threads.

> 16-core workstation

I meant to write 6-core here.

**Comment by jcupitt** [comment]
> I meant to write 6-core here.

Ah, I was impressed!

I've got a branch which gets systime back down to 2m on that test, it just needs a little polish still.

How about something to size threadpools by pipeline complexity? 

We'd add a new class member to `VipsOperation` called eg. `parallelism` which is an estimate of how much the operation would benefit from threading. So `gaussblur` would be 10 (perhaps) and `jpegsave` might stay at the default of 0.

When building output images, we'd take the max of an image metadata item called `parallelism` on the inputs, add the value for this operator, and set that on the output image.

In the various sinks, we'd use the value of `parallelism` on the image being iterated over to set the threadpool size.

`dzsave` would see a huge improvement since the tile saver would run with a single thread and we wouldn't get this x ^ 2 expansion in thread numbers.

**Comment by kleisauke** [comment]
The [`reimplement-threadpool`](https://github.com/libvips/libvips/tree/reimplement-threadpool) branch is somewhat slower here, and doesn't seem to reuse threads (making it vulnerable to CPU oversubscription problems under heavy server load).
```bash
$ time vips dzsave CMU-1.svs x

real	0m38.415s
user	1m55.072s
sys	0m45.563s
```

I agree, something to size the threadpools by pipeline complexity would help `dzsave`.

**Comment by kleisauke** [comment]
As an aside, I was recently made aware of https://github.com/taskflow/taskflow via https://www.youtube.com/watch?v=MX15huP5DsM, which seems interesting for use in libvips.

**Comment by jcupitt** [comment]
Yes, I’ve not hooked up the thread reuser yet, but systime is down from 20m
to 2m on dzsave. I’ll try to finish it tonight.

**Comment by jcupitt** [comment]
I now see:

```
$ time vips dzsave CMU-1.svs x
real	1m24.414s
user	7m12.531s
sys	28m42.812s
```

I was baffled by the huge systime -- how could this thread recycler be *slower* than GThreadpool (which is much more complex). But of course high systime means the 1,000+ threads in the pool are spending a lot of time waiting on condition variables! So of cource a thread recycler will have much higher systime than the old start/stop system. WHat an idiot.

I'll try hacking up a thing to size threadpools by pipeline complexity. It should prevent overcommitting and bring the time right down.

**Comment by jcupitt** [comment]
I added a few more features:

- threadpools now size dynamically. They start off at concurrency / 2 (ie. half size), then scale up or down depending on how many workers are stalled or processing. If all workers are busy then pool is too small, if many are blocked then it's too big. This stops threadpool committing many more threads than are necessary.
- I added a `concurrency` image metadata item that anyone can set that sets the max threadpool size for operations on that image. `dzsave` uses this to limit the concurrency on tile save.
- I tweaked the thing that picks the tile size for iteration. It now only selects `thinstrip` for very wide images, so you don't get tiles with only a few 100 pixels in.

I now see:

git master:

```
$ time vips dzsave CMU-1.svs x
real	0m47.673s
user	4m21.642s
sys	13m11.011s
```

And `reimplement-threadpool`:

```
$ time vips dzsave CMU-1.svs x --vips-leak
real	0m23.578s
user	2m18.459s
sys	1m56.690s
```

So it looks promising! Peak number of threads is down from 1000+ to under 200.

Sadly although `dzsave` seems to work, it coredumps on the test suite :( I'm not sure why.

**Comment by jcupitt** [comment]
vips-bench performance would need checking too, this branch loses a bit there right now.

**Comment by jcupitt** [comment]
I did a final round of fixing and tuning. The test suite now passes and I see:

```
$ time vips dzsave CMU-1.svs[rgb] x
real	0m13.706s
user	1m34.806s
sys	1m13.540s
```

I'll check vips-bench performance.

**Comment by kleisauke** [comment]
Great! I also see a nice speed-up here.

```bash
$ time vips dzsave CMU-1.svs[rgb] x

real	0m20.909s
user	1m6.279s
sys	0m11.938s
```

I will run some more benchmarks later this week. wasm-vips also needs some fixes after this. It currently requires an exclusive threadpool where a fixed number of threads are started during `vips_init` (see e.g. [this patch](https://github.com/kleisauke/wasm-vips/blob/master/build/patches/vips-1492-emscripten.patch)), since `pthread_create()` calls require returning to the main event loop, ouch.

BTW, do you think it still makes sense that the `->start` function is outside the `pool->allocate_lock`? See e.g.  commit https://github.com/kleisauke/libvips/commit/1d869af27ee6f40877e91fd7f5ba85f32f20498d.

**Comment by jcupitt** [comment]
`vips-bench` is only a little slower than master, but quite a bit slower than 8.11:

Version | dzsave | vips-c
---- | ---- | ----
8.11 | 44s | 0.37s
master | 48s |  0.57s
this | 20s (14s in rgb mode) | 0.63s

- could be startup time
- could be only for high core numbers, I'll try on a two core laptop
- at least dzsave is quicker heh

**Comment by jcupitt** [comment]
> wasm-vips also needs some fixes after this. 

Huh interesting. We could add a cap to the new threadset thing and use that for all thread operations (including render etc.). Peak thread counts are way down with this branch: the dynamic pool sizing really helps, especially with dzsave. I see 120 threads peak now, that's 10x less. The peak would be a lot lower again with fewer cores.

How do you detect the number of cores for wasm-vips? If it's only 4 or so, capping threadset at 20 should give plenty of room.

**Comment by jcupitt** [comment]
> BTW, do you think it still makes sense that the ->start function is outside the pool->allocate_lock? 

I think that was just by analogy with the start/generate/stop callbacks for regions. They have locks for start and stop, so I think I thought threadpool should do that too. Do you think removing these locks makes a useful difference to performance?

**Comment by jcupitt** [comment]
I tried a few more things and it seems this branch is roughly equal to master on vips-bench across the system I have available for test. I think it's probably OK to merge, in terms of performance, at least.

Both are slower than 8.11 by a fairly large percentage, we should investigate.

**Comment by kleisauke** [comment]
> How do you detect the number of cores for wasm-vips? If it's only 4 or so, capping threadset at 20 should give plenty of room.

It's currently limited to [`navigator.hardwareConcurrency + 3`](https://github.com/kleisauke/wasm-vips/blob/ec44ecd6fbc7b3da7c6a9f71a8989cd6b06043a4/src/CMakeLists.txt#L118-L119). The +3 is the amount of additional background threads that may run within the threadpool.

Don't spend too much time one this though, it's [non-POSIX behavior](https://emscripten.org/docs/porting/pthreads.html#special-considerations), `pthread_create()` and immediately `pthread_join()` or synchronously waiting for the thread to become available will cause a surprising deadlock. Hence, the threadpool is not dynamically sized in wasm-vips.

> Do you think removing these locks makes a useful difference to performance?

Not sure if this can be measured, so perhaps it doesn't matter. I had noticed, at that time, that most `_start()` functions just use it to initialize an accumulator, so it can be called concurrently. See for example:
https://github.com/libvips/libvips/blob/f11b58ff6c602549d553c57e506142de5b6d28e9/libvips/arithmetic/avg.c#L113-L120
https://github.com/libvips/libvips/blob/f11b58ff6c602549d553c57e506142de5b6d28e9/libvips/iofuncs/generate.c#L441-L447

Though, for things like this:
https://github.com/libvips/libvips/blob/f11b58ff6c602549d553c57e506142de5b6d28e9/libvips/arithmetic/profile.c#L170-L173

It still needs to be guarded with a lock.

Note that `sink->sslock` can be safely removed on the [`reimplement-threadpool`](https://github.com/libvips/libvips/tree/reimplement-threadpool) branch now, see e.g. commit https://github.com/kleisauke/libvips/commit/4e62c0d943480c0563425c324e2cf93af2ec877c.

> I think it's probably OK to merge, in terms of performance, at least.

Feel free to merge.

> Both are slower than 8.11 by a fairly large percentage, we should investigate.

Is this also the case on 8.12? I remembered that the introduction of `vips_tiffsave_target()` in 8.13 caused a performance regression on the NetVips benchmarks (83.10ms -> 114.4ms, see commit https://github.com/kleisauke/net-vips/commit/707f693fe40816f479bba516e2ce13c8f74b905e).

**Comment by jcupitt** [comment]
Yes, 8.12 is relatively slow too, it's why I've not updated the "speed and memory use" page for a while.

**Comment by jcupitt** [comment]
OK, let's merge, and improve wasm-vips compatibility and revise performance in another PR or two.
